arch/cortex-m/src/lib.rs

```diff
@@ -243,14 +249,21 @@ pub unsafe extern "C" fn switch_to_user_arm_v7m(
 
     // Push non-hardware-stacked registers into the saved state for the
     // application.
-    stmia $2, {r4-r11}
+    stmia r1, {{r4-r11}}
 
     // Update the user stack pointer with the current value after the
     // application has executed.
-    mrs $0, PSP   // r0 = PSP"
-    : "={r0}"(user_stack)
-    : "{r0}"(user_stack), "{r1}"(process_regs)
-    : "r4","r5","r6","r8","r9","r10","r11" : "volatile" );
+    mrs r0, PSP   // r0 = PSP
+
+    // Need to restore r6 since we clobbered it when switching to and from the
+    // app.
+    mov r6, r3
+    ",
+    inout("r0") user_stack,
+    in("r1") process_regs,
+    out("r3") _, out("r4") _, out("r5") _, out("r8") _, out("r9") _,
```

Comment:
It wasn't here before either.., but why is there no `r7`?

Reply:
I believe r7 is for the frame pointer, and returns the same error as trying to mark r6 as a clobber.

Reply:
Do we need to be doing the same thing (under the same premise that userspace can/will likely clobber it)?

Reply:
I would say no, since we haven't been so far with no negative consequences.

Reply:
https://github.com/rust-lang/rust/issues/81092#issuecomment-784524675 suggests that it might be safest to do so, though I did not ask explicitly about `r7`

Reply:
That feels more like we might be getting lucky the being correct. Assuming nothing has changed since my compilers class project in 2012, all else being equal, llvm will allocate registers following a lowest-number-first policy. In practice, it's likely `r7` simply hasn't been live in a meaningful way for kernel code across syscalls, but that doesn't it couldn't be. We should probably save `r7` as well.

Reply:
I don't think we have just been getting lucky, but I think it makes sense to be defensive since we have another register (r2) to spare. I added a commit which saves r7.